private/Metaprogram.h

```diff
@@ -0,0 +1,349 @@
+   constexpr auto expectedSize = 2;
+
+   // When
+   using Candidates = extract_candidates<SampleTypes, Dependencies, Path>::type;
+
+   // Then
+   static_assert(expectedSize == Candidates::size());
+   static_assert(std::is_same<char, type_list_type_at<0, Candidates>::type >::value);
+   static_assert(std::is_same<float, type_list_type_at<1, Candidates>::type >::value);
+}
+
+namespace extract_candidates_for_nodes_without_dependencies_and_partially_full_path
+{
+   // Given
+   using SampleTypes = type_list<long, char, double, float, int>;
+   using Dependencies = type_list<type_list<>, type_list<>, type_list<>, type_list<>, type_list<>>;  
+   using Path = type_list<double, int>;
+   constexpr auto expectedSize = 3;
+
+   // When
+   using Candidates = extract_candidates<SampleTypes, Dependencies, Path>::type;
+
+   // Then
+   static_assert(expectedSize == Candidates::size());
+   static_assert(std::is_same<long, type_list_type_at<0, Candidates>::type >::value);
+   static_assert(std::is_same<char, type_list_type_at<1, Candidates>::type >::value);
+   static_assert(std::is_same<float, type_list_type_at<2, Candidates>::type >::value);
+}
+
+namespace sequential_extraction_of_candidates
+{
+   // Given
+   // We repeat the extraction steps with an empty path and some dependencies
+   using SampleTypes = type_list<long, char, double, float, int>;
+   using Dependencies = type_list<type_list<int,char>, type_list<>, type_list<float>, type_list<>, type_list<char>>;  
+   using Path = type_list<>;
+   using FirstPassCandidates = extract_candidates<SampleTypes, Dependencies, Path>::type;
+
+   // When
+   // We move the candidates from the first pass to the path (so it becomes the new Path)
+   using SecondPassPath = FirstPassCandidates;
+   using SecondPassCandidates = extract_candidates<SampleTypes, Dependencies, SecondPassPath>::type;
+   // Since we already have char and float on the path, int and double can be added to
+   // the path, but not yet long, because it depends on int
+   constexpr auto expectedSizeSecondPass = 2;
+
+   // Then
+   static_assert(expectedSizeSecondPass == SecondPassCandidates::size());
+   static_assert(std::is_same<double, type_list_type_at<0, SecondPassCandidates>::type >::value);
+   static_assert(std::is_same<int, type_list_type_at<1, SecondPassCandidates>::type >::value);
+
+   // When
+   using ThirdPassPath = type_list_concat_lists<FirstPassCandidates, SecondPassCandidates>::type;
+   using ThirdPassCandidates = extract_candidates<SampleTypes, Dependencies, ThirdPassPath>::type;
+   constexpr auto expectedSizeThirdPass = 1;
+   static_assert(expectedSizeThirdPass == ThirdPassCandidates::size());
+   static_assert(std::is_same<long, type_list_type_at<0, ThirdPassCandidates>::type >::value);
+}
+
+namespace full_topological_sort
+{
+   // Given
+   using SampleTypes = type_list<long, char, double, float, int>;
+   // Arbitrarily, let's say that double depends on float, int depends on char,
```

Comment:
Light has no dependencies
Air has no dependencies
Water depends on Light and Air
Soil depends on Light, Air and Water
Flower depends on Soil